compiler/src/iree/compiler/Codegen/Utils/GPUUtils.cpp

```diff
@@ -438,9 +439,26 @@ static Value warpReduction(Location loc, OpBuilder &builder, Value input,
   const bool needsPacking = kShuffleBitWidth != origBitWidth;
   IntegerType equivIntType = builder.getIntegerType(origBitWidth);
 
-  // Always perform the shuffles over the supported scalar type. For inputs of
-  // smaller bitwidth, perform packing and unpacking via the supported integer
-  // type.
+  // Defer expansion of subgroup reduction until later in pass pipeline to
+  // enable conditional lowering to DPP ops, for potential perf gains over
+  // gpu.shuffle ops.
+  if (!expandSubgroupReduce && numLaneToReduce <= warpSize &&
+      warpSize % numLaneToReduce == 0) {
+    auto gpuReduceKind = combiningKindToAllReduce(kind);
```

Comment:
Could we spell out the exact type here? It's not obvious to me based on the RHS. See https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable.